File: app/src/main/java/org/deiverbum/app/data/entity/TodayTercia.java
package org.deiverbum.app.data.entity;

import androidx.room.Embedded;
import androidx.room.Relation;

import org.deiverbum.app.model.BiblicaBreve;
import org.deiverbum.app.model.Himno;
import org.deiverbum.app.model.Hoy;
import org.deiverbum.app.model.Intermedia;
import org.deiverbum.app.model.Oracion;
import org.deiverbum.app.model.Salmodia;
import org.deiverbum.app.model.Santo;

import java.util.List;

/**
 * @author A. Cedano
 * @version 1.0
 * @since 2022.2
 */
public class TodayTercia {

    @Embedded
    public Today today;

    @Relation(
            entity = SaintEntity.class,
            parentColumn = "saintFK",
            entityColumn = "saintID" //liturgiaId
    )
    public SaintEntity santo;


    @Relation(
            entity = LHHymnJoinEntity.class,
            parentColumn = "tHymnFK",
            entityColumn = "groupID"
    )
    public LHHymnWithAll himno;

    @Relation(
            entity = LHReadingShortJoinEntity.class,
            parentColumn = "tBiblicalFK",
            entityColumn = "groupID"
    )
    public LHReadingShortAll biblica;

    @Relation(
            entity = LHPsalmodyJoinEntity.class,
            parentColumn = "tPsalmodyFK",
            entityColumn = "groupID"
    )
    public LHPsalmody salmodia;

    @Relation(
            entity = PsalmodyEntity.class,
            parentColumn = "tPsalmodyFK",
            entityColumn = "groupFK"
    )
    public List<PsalmodyWithPsalms> salmos;

    @Relation(
            entity = LHPrayerEntity.class,
            parentColumn = "tPrayerFK",
            entityColumn = "groupID"
    )
    public LHPrayerAll lhPrayerAll;



    @Relation(
            entity = LiturgyEntity.class,
            parentColumn = "liturgyFK",
            entityColumn = "liturgyID"
    )
    public LiturgyWithTime feria;

    @Relation(
            entity = LiturgyEntity.class,
            parentColumn = "previousFK",
            entityColumn = "liturgyID"
    )
    public LiturgyWithTime previo;


    public Hoy getToday(){
        Hoy dm = new Hoy();
        dm.setFeria(feria.getDomainModel());
        dm.setFecha(String.valueOf(today.getHoy()));
        dm.setCalendarTime(today.tiempoId);
        dm.setHasSaint(true);
        //dm.setMLecturasFK(today.mLecturasFK);
        //dm.setPrevio(previo.getDomainModel());
        dm.setTitulo(feria.getDomainModel().getNombre());
        return dm;
    }

    public Intermedia getDomainModel(){
        Intermedia dm=new Intermedia();
        dm.setHoy(getToday());
        dm.setHourId(3);
        dm.setHimno(getHimno());
        dm.setSalmodia(getSalmodia());
        dm.setLecturaBreve(getBiblica());
        dm.setOracion(getOracion());
        return dm;
    }

    public Himno getHimno(){
        return himno.getDomainModel();
    }



    public BiblicaBreve getBiblica(){
        return  biblica.getDomainModel(today.getTiempoId());
    }



    public Santo getSanto(){
        return  santo.getDomainModel(false);
    }


    public Salmodia getSalmodia() {

        return salmodia.getDomainModel();
    }

    public Oracion getOracion() {
        return lhPrayerAll.getDomainModel();
    }


}

File: app/src/main/java/org/deiverbum/app/data/entity/relation/TodayTercia.kt
package org.deiverbum.app.data.entity.relation

import androidx.room.Embedded
import androidx.room.Relation
import org.deiverbum.app.data.entity.*
import org.deiverbum.app.model.*

/**
 * Obtiene el contenido de la hora **`Tercia`** desde la base de datos.
 *
 * @author A. Cedano
 * @version 1.0
 * @since 2023.1
 */
data class TodayTercia (

    @Embedded
    var today: TodayEntity? = null,

    @Relation(entity = LHHymnJoinEntity::class, parentColumn = "tHymnFK", entityColumn = "groupID")
    var himno: LHHymnWithAll? = null,

    @Relation(
        entity = LHReadingShortJoinEntity::class,
        parentColumn = "tBiblicalFK",
        entityColumn = "groupID"
    )
    var biblica: LHReadingShortAll? = null,

    @Relation(
        entity = LHPsalmodyJoinEntity::class,
        parentColumn = "tPsalmodyFK",
        entityColumn = "groupID"
    )
    var salmodia: LHPsalmodyAll? = null,

    @Relation(
        entity = LHPsalmodyEntity::class,
        parentColumn = "tPsalmodyFK",
        entityColumn = "groupFK"
    )
    var salmos: List<PsalmodyWithPsalms>? = null,

    @Relation(entity = LHPrayerEntity::class, parentColumn = "tPrayerFK", entityColumn = "groupID")
    var lhPrayerAll: LHPrayerAll? = null,

    @Relation(entity = LiturgyEntity::class, parentColumn = "liturgyFK", entityColumn = "liturgyID")
    var feria: LiturgyWithTime = LiturgyWithTime(),

    @Relation(
        entity = LiturgyEntity::class,
        parentColumn = "previousFK",
        entityColumn = "liturgyID"
    )
    var previo: LiturgyWithTime? = null)
{
    fun getToday(): Today {
        val dm = Today()
        dm.liturgyDay = feria.domainModel
        dm.liturgyPrevious = if (today!!.previoId > 1) previo?.domainModel else null
        dm.todayDate = today!!.hoy
        return dm
    }

    val domainModel: Liturgy
        get() {
            val dm = feria.domainModel
            val bh = BreviaryHour()
            val hi = Intermedia()
            dm.typeID = 3
            dm.today = getToday()
            hi.today = getToday()
            hi.setHimno(getHimno())
            hi.setSalmodia(getSalmodia())
            hi.lecturaBreve = getBiblica()
            hi.setOracion(oracion)
            bh.setIntermedia(hi)
            dm.breviaryHour = bh
            return dm
        }

    private fun getHimno(): LHHymn? {
        return himno?.domainModel
    }

    private fun getBiblica(): BiblicalShort {
        return biblica!!.getDomainModel(today!!.tiempoId)
    }

    private fun getSalmodia(): LHPsalmody? {
        return salmodia?.domainModel
    }

    val oracion: Prayer?
        get() = lhPrayerAll?.domainModel
    val domainModelToday: Today
        get() {
            val dm = feria.domainModel
            val bh = BreviaryHour()
            val hi = Intermedia()
            val dmToday = getToday()
            dm.typeID = 3
            hi.setTypeId(3)
            hi.setHimno(getHimno())
            hi.setSalmodia(getSalmodia())
            hi.lecturaBreve = getBiblica()
            hi.setOracion(oracion)
            bh.setIntermedia(hi)
            dm.breviaryHour = bh
            dmToday.liturgyDay = dm
            return dmToday
        }
}